feature/esp32c6: Support for esp32c6

raffarost opened the pull request:
Initial support for esp32c6

- Basic drivers and peripherals 
- Basic DT configuration
- Support for esp32c6_devkitc board, hwmv2
- Simple boot and MCUboot support

LucasTambor committed:
dt-bindings: add esp32c6 signals

Add clock, interrupt controller and pinctrl related macros
for esp32c6

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
dts: riscv: espressif: add esp32c6

Add esp32c6 basic device tree.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
boards: espressif: esp32c6: add esp32c6_devkitc

Add esp32c6_devkitc basic support.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
drivers: Update to add support for esp32c6

Changes to bring support for esp32c6 SoC.
- clock control
- gpio
- pinctrl
- serial
- timer

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

Commit:
drivers: watchdog: Added support to C6

Added support to watchdog timer to ESP32C6

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
west.yml: update for esp32c6 support

hal_espressif update for C6 support

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

nordicjm commented on soc/espressif/esp32c6/Kconfig:
```diff
@@ -0,0 +1,47 @@
+config MAC_BB_PD
+	bool "Power down MAC and baseband of Wi-Fi and Bluetooth when PHY is disabled"
+	depends on SOC_SERIES_ESP32C6 && TICKLESS_KERNEL
+	default n
```
```suggestion
```

nordicjm replied:
this is not fixed

raffarost replied:
Can you check again? I see this as fixed already

sylvioalves commented:
@nordicjm still a draft version, not fully reviewed.. we are working on it.

rftafas commented:
> @nordicjm still a draft version, not fully reviewed.. we are working on it.

For the record, we anticipated it to help the efforts of those discussing it in https://github.com/zephyrproject-rtos/zephyr/discussions/62138

Commit:
drivers: spi: Add suport to ESP32C6

Added GP-SPI2 (general purpose SPI2) support for ESP32C6

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
boards: esp32c6_devkitc: Added support for GP-SPI2

Added support for general purpose SPI2 (GP-SPI2)

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

ajf58 commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -0,0 +1,264 @@
+Developers can either connect peripherals with jumper wires or mount ESP32-C6-DevKitC-1 on
+a breadboard. [2]_
+
+ESP32-C6 includes the fallowing features:
```
nit: typo.
```suggestion
ESP32-C6 includes the following features:
```

ajf58 commented on soc/espressif/esp32c6/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+# ESP32C3 board configuration
```
nit: a little typo here from the copying across.

ajf58 commented on soc/espressif/esp32c6/mcuboot.ld:
```diff
@@ -0,0 +1,287 @@
+    *(.iram1 .iram1.*)
+    *(.iram0.literal .iram.literal .iram.text.literal .iram0.text .iram.text)
+
+    /* C3 memprot requires 512 B alignment for split lines */
```
nit: a little typo here

Commit:
drivers: intc: Fix for ESP32C6 interrupt sources allocation

Fix to properly allocate IRQs for interrupt sources over 60.
It also screens out non-allocatable IRQs used by the CPU.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

marekmatej committed:
soc: espressif: add esp32-c6 support

Add basic support for esp32c6 SoC.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

Commit:
drivers: ledc: Clock source update to support ESP32C6

Clock source SCLK added for C6 on LEDC

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
boards: esp32c6: Added PWM test file for esp32c6_devkitc

Added basic PWM test for esp32c6_devkitc

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
drivers: clock_control: Refactor for ESP32C6

Added support for C6 to allow CPU clock config

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
tests: rtc_clk: Added support to ESP32C6

Added support for C6 for CPU and LP clock testing

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
soc: esp32c6: Kconfig and .ld updates, DTS and comments fix

Kconfig, .ld and comments fixing
Fixed address of UART1, WDT and RTC timer disabled by default

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

sylvioalves commented on boards/espressif/esp32c6_devkitc/Kconfig.defconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Espressif Systems (Shanghai) Co., Ltd.
+# SPDX-License-Identifier: Apache-2.0
+
+config HEAP_MEM_POOL_SIZE
```
Please, update this to:
```c
config HEAP_MEM_POOL_ADD_SIZE_BOARD
	int
	default 65535 if WIFI && BT
	default 51200 if WIFI
	default 40960 if BT
	default 4096
```

sylvioalves commented on boards/espressif/esp32c6_devkitc/Kconfig.esp32c6_devkitc:
```diff
@@ -0,0 +1,7 @@
+# ESP32C6 devkitc board configuration
+
+# Copyright (c) 2023 Espressif Systems (Shanghai) Co., Ltd.
```
Use 2024 instead.

sylvioalves commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -0,0 +1,264 @@
+Overview
+********
+
+ESP32-C6 is Espressif’s first Wi-Fi 6 SoC integrating 2.4 GHz Wi-Fi 6, Bluetooth 5.3 (LE) and the
```
typo: `Espressif's`

raffarost replied:
Is that a typo? The possessive form "Espressif’s" means that ESP32-C6 is a product of Espressif

sylvioalves commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -0,0 +1,264 @@
++------------+------------+-------------------------------------+
+| USB-JTAG   | on-chip    | hardware interface                  |
++------------+------------+-------------------------------------+
+
```
In case ledc/pwm, interrupt, spi, pinctrl, timer, watchdog and are supported, please, include it here as well.

sylvioalves replied:
@raffarost, please, update this entry as needed.

sylvioalves commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -0,0 +1,264 @@
+Building & Flashing
+*******************
+
+ESP-IDF bootloader
```
There is no longer esp-idf bootloader. Update accordingly. You can chec other esp32-based board.

sylvioalves commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc.yaml:
```diff
@@ -0,0 +1,14 @@
+toolchain:
+  - zephyr
+supported:
+  - gpio
```
If you think there are additional peripherals already supported, add it here. Consider checking additional overlays in samples/test folder as well if needed.

sylvioalves commented on drivers/clock_control/clock_control_esp32.c:
```diff
@@ -32,6 +32,17 @@
+#include <esp32c6/rom/rtc.h>
+#include <soc/dport_access.h>
+#include <hal/clk_tree_ll.h>
+#undef MHZ
```
This sounds wrong. Can this be removed and reworked?

raffarost replied:
This was discussed with Lucas. Not a nice workaround but otherwise would require changing many HAL files, thus affecting other SoC's.

gmarull replied:
Please, fix HAL, this is not allowed per guidelines.

sylvioalves commented on drivers/interrupt_controller/intc_esp32c3.c:
```diff
@@ -75,6 +129,18 @@ void esp_intr_initialize(void)
 		esp_rom_intr_matrix_set(0, i, ESP32C3_INTC_DISABLED_SLOT);
 	}
```
Also, I see a few ESP32C3_xx related definitions. Can you confirm it matches C6 values?

raffarost replied:
It is correct. The driver will be reworked in a next version for a generic implementation (issue is opened).

sylvioalves commented on soc/espressif/esp32c6/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	select PINCTRL
+	select RISCV_ISA_RV32I
+	select RISCV_ISA_EXT_M
+	select RISCV_ISA_EXT_C
```
ESP32-C6 also support Atomic operations. Please add `select RISCV_ISA_EXT_A`.

sylvioalves commented on dts/riscv/espressif/esp32c6/esp32c6_common.dtsi:
```diff
@@ -0,0 +1,170 @@
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "espressif,riscv";
+			riscv,isa = "rv32imc_zicsr";
```
It support atomc operation. Could be `rv32imac_zicsr` I think?

sylvioalves commented on soc/espressif/esp32c6/Kconfig:
```diff
@@ -0,0 +1,47 @@
+
+if SOC_SERIES_ESP32C6
+
+config IDF_TARGET_ESP32C6
+	bool "ESP32C6 as target board"
+	default y
```
This symbol is used in HAL layer. I think it can be hidden though.

nordicjm replied:
ping

sylvioalves commented on soc/espressif/esp32c6/Kconfig.mac:
```diff
@@ -0,0 +1,48 @@
+
+if SOC_SERIES_ESP32C6
+
+choice ESP32C6_UNIVERSAL_MAC_ADDRESSES
```
This is shared between Wi-Fi and Bluetooth. Other ESP32 SoC already embed this in soc level.

sylvioalves commented on soc/espressif/esp32c6/Kconfig.soc:
```diff
@@ -0,0 +1,37 @@
+	default "esp32c6" if SOC_SERIES_ESP32C6
+
+config SOC_PART_NUMBER
+	default "ESP32_C6_WROOM_1U_N4" if SOC_ESP32_C6_WROOM_1U_N4
+	default "ESP32_C6_WROOM_1U_N8" if SOC_ESP32_C6_WROOM_1U_N8
+	default "ESP32C6" if SOC_ESP32C6
```
`default "ESP32C6" if SOC_ESP32C6` should be removed.

sylvioalves commented on soc/espressif/esp32c6/default.ld:
```diff
@@ -0,0 +1,811 @@
+
+    *libzephyr.a:cpu_region_protect.*(.literal .text .literal.* .text.*)
+
+    /* TODO: optimise */
```
Please, remove this comment.

sylvioalves commented on soc/espressif/esp32c6/default.ld:
```diff
@@ -0,0 +1,811 @@
+    _bss_start = ABSOLUTE(.);
+
+    /*  bluetooth library requires this symbol to be defined */
+    _btdm_bss_start = ABSOLUTE(.);
```
Is this BT entry necessary?

raffarost replied:
Removed

sylvioalves commented on west.yml:
```diff
@@ -157,7 +157,7 @@ manifest:
         - hal
     - name: hal_espressif
-      revision: b5f5fa6d08780c0077407150c6685d6c9d70262c
+      revision: e67c3da707c8ce1185177a0a357259fb89498cf4
```
Update this when hal_espressif PR is merged.

sylvioalves commented on west.yml:
```diff
@@ -287,7 +287,7 @@ manifest:
         - crypto
     - name: mcuboot
-      revision: 018dc01d484358d81f889c36471d9ce5590965b6
+      revision: 898a1ca64a759541d9fcc37fef921db93d99ad70
```
mcuboot update could be done separately. Restore this to original value.

raffarost commented on soc/espressif/esp32c6/Kconfig.rtc:
```diff
@@ -0,0 +1,50 @@
+
+if SOC_SERIES_ESP32C6
+
+choice RTC_CLK_SRC
```
File was properly removed with clock control update commit

raffarost commented on dts/riscv/espressif/esp32c6/esp32c6_common.dtsi:
```diff
@@ -0,0 +1,143 @@
+			status = "okay";
+		};
+
+		rtc: rtc@600b000 {
```
RTC alone in some chips will refer to a clock control system, while 'RTC timer' will stand for the timer you might refer to. RTC must stay enabled, while 'RTC timer' can be disabled indeed

nordicjm replied:
so why is rtc timer enabled if it's not enabled (below this point)?

raffarost replied:
RTC timer was properly disabled, while RTC will still remain enabled as it is necessary for clock configuration during startup

raffarost commented on drivers/serial/serial_esp32_usb.c:
```diff
@@ -22,7 +22,7 @@
 
-#ifdef CONFIG_SOC_SERIES_ESP32C3
+#if defined(CONFIG_SOC_SERIES_ESP32C3) || defined(CONFIG_SOC_SERIES_ESP32C6)
 #define ISR_HANDLER isr_handler_t
```
We'll improve those when there's a driver rework

gmarull reviewed (changes requested):
Fix HAL definition colliding with Zephyr

gmarull reviewed (dismissed):
please, organize commits properly and add accurate descriptions:

```
dt-bindings: add esp32c6 signals
Add clock, interrupt controller and pinctrl related macros
for esp32c6

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

pinctrl: fix for CI (C6)

Fix for CI

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>
```
?

nordicjm commented on soc/espressif/esp32c6/CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+        ${PYTHON_EXECUTABLE} ${ESP_IDF_PATH}/tools/ci/check_callgraph.py
+      ARGS
+        --rtl-dirs ${CMAKE_BINARY_DIR}/zephyr
+	--elf-file ${CMAKE_BINARY_DIR}/zephyr/${CONFIG_KERNEL_BIN_NAME}.elf
+        find-refs --from-section=.iram0.iram_loader --to-section=.iram0.text
```
indent looks off

nordicjm commented on soc/espressif/esp32c6/Kconfig.soc:
```diff
@@ -0,0 +1,29 @@
+config SOC_PART_NUMBER
+	default "ESP32_C6_WROOM_1U_N4" if SOC_ESP32_C6_WROOM_1U_N4
+	default "ESP32_C6_WROOM_1U_N8" if SOC_ESP32_C6_WROOM_1U_N8
+	default "ESP32C6" if SOC_ESP32C6
```
I don't see how this can ever be selected, boards will select `SOC_ESP32_C6_WROOM_1U_N4` or `SOC_ESP32_C6_WROOM_1U_N8` right?

nordicjm commented on soc/espressif/esp32c6/Kconfig.soc:
```diff
@@ -0,0 +1,29 @@
+	default "esp32c6" if SOC_SERIES_ESP32C6
+
+config SOC
+	default "esp32c6" if SOC_SERIES_ESP32C6
```
`... if SOC_ESP32C6`

nordicjm commented on drivers/interrupt_controller/Kconfig.esp32:
```diff
@@ -5,7 +5,7 @@
 	bool "Interrupt allocator for Xtensa-based Espressif SoCs"
-	default y if SOC_SERIES_ESP32
+	default y if SOC_FAMILY_ESPRESSIF_ESP32 && !SOC_SERIES_ESP32C3 && !SOC_SERIES_ESP32C6
 	help
```
you changed this **to** `SOC_SERIES_ESP32` in an earlier commit in this PR and change it again, so it should be fixed in the original commit instead of changing it to something wrong then fixing it

nordicjm commented on soc/espressif/esp32c6/Kconfig.rtc:
```diff
@@ -1,50 +0,0 @@
-
-if SOC_SERIES_ESP32C6
-
-choice RTC_CLK_SRC
```
what is going on in this pr!? You add this file in one commit then remove it in another. Fix this pr

sylvioalves commented on scripts/ci/check_compliance.py:
```diff
@@ -870,6 +870,8 @@ def check_no_undef_outside_kconfig(self, kconf):
         "CLOCK_STM32_SYSCLK_SRC_",
         "CMU",
         "COMPILER_RT_RTLIB",
+        "ULP_COPROC_ENABLED",		# Added when coprocessor is supported
```
Where is it defined that causes CI compliance issue?

sylvioalves commented on soc/espressif/esp32c6/memory.h:
```diff
@@ -0,0 +1,63 @@
+/* These lengths can be adjusted, if necessary: */
+#define BOOTLOADER_DRAM_SEG_LEN        0x9000
+#define BOOTLOADER_IRAM_LOADER_SEG_LEN 0x3000
+#define BOOTLOADER_IRAM_SEG_LEN        0x8000
```
@raffarost @marekmatej Do we need to increase IRAM_SEG_LEN due to mcuboot updates?

marekmatej replied:
The memory layout could be improved. I'll add the commit.

str4t0m reviewed (dismissed):
Please don't use commit hashes of unmerged HAL PRs in the west.yaml. Use  `pull/278/head` while in review. Then merge the HAL PR, and finally use the correct SHA. Otherwise we might end up merging a PR with an invalid hal reference.

raffarost commented:
> Please don't use commit hashes of unmerged HAL PRs in the west.yaml. Use `pull/278/head` while in review. Then merge the HAL PR, and finally use the correct SHA. Otherwise we might end up merging a PR with an invalid hal reference.

Thanks for your comment

sylvioalves commented on soc/espressif/esp32c6/default.ld:
```diff
@@ -0,0 +1,807 @@
+  /* --- START OF .rodata --- */
+
+  /* Align next section to 64k to allow mapping */
+  .flash.align_rodata (NOLOAD) :
+  {
+    /* Subsequent segment lma and vma align */
+    . = ALIGN(CACHE_ALIGN);
+  } GROUP_DATA_LINK_IN(CACHED_REGION, ROMABLE_REGION)
```
`.flash.align_rodata` is duplicated. Remove this.

sylvioalves commented on soc/espressif/esp32c6/default.ld:
```diff
@@ -0,0 +1,807 @@
+    . = ALIGN(CACHE_ALIGN);
+  } GROUP_DATA_LINK_IN(CACHED_REGION, ROMABLE_REGION)
+
+  .flash.align_rom (NOLOAD) :
```
I think it needs to account for the .text region:
````c
  .flash.align_rom (NOLOAD) :
  {
    /* Spacer in the IROM address to avoid interfering with the DROM address
     * because DROM and IROM regions share the same address space */
    . += SIZEOF(.flash.text);
    . = ALIGN(CACHE_ALIGN);
  } GROUP_LINK_IN(CACHEd_REGION)
  ```

sylvioalves replied:
@marekmatej PTAL as well.

marekmatej replied:
please, update to
```
  /* Align next section to 64k to allow mapping */
  .flash.align_rom (NOLOAD) :
  {
    . = ALIGN(CACHE_ALIGN);
  } GROUP_LINK_IN(ROMABLE_REGION)

  .flash.align_rodata (NOLOAD) :
  {
    . = ALIGN(CACHE_ALIGN);
  } GROUP_LINK_IN(CACHED_REGION)
```

sylvioalves commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -0,0 +1,275 @@
++------------+------------+-------------------------------------+
+| Interface  | Controller | Driver/Component                    |
++============+============+=====================================+
++------------+------------+-------------------------------------+
```
Remove this extra line:
`+------------+------------+-------------------------------------+`

marekmatej commented on west.yml:
```diff
@@ -157,7 +157,7 @@ manifest:
         - hal
     - name: hal_espressif
-      revision: 0690c03f1540e8a5082da4c2b997e64bde8f4765
+      revision: pull/278/head
```
update the `mcuboot` to the `898a1ca64a759541d9fcc37fef921db93d99ad70` which has C6 support.

sylvioalves commented:
Also tested it locally:
` Total complete: 2701/2701  100%  skipped: 2273, failed:    0, error:    0`

marekmatej commented:
C6 related changes https://github.com/zephyrproject-rtos/zephyr/pull/74320

kartben reviewed (dismissed):
fwiw I was able to get blinky running on https://docs.m5stack.com/en/core/M5NanoC6 🌟 🚀 
+1 for the docs aspects as well

nordicjm commented on soc/espressif/esp32c6/Kconfig.soc:
```diff
@@ -0,0 +1,28 @@
+	select SOC_SERIES_ESP32C6
+
+config SOC_SERIES
+	default "esp32c6" if SOC_ESP32C6
```
`	default "esp32c6" if SOC_SERIES_ESP32C6`

sylvioalves replied:
Nice catch!

kartben commented:
@raffarost @sylvioalves needs rebase

kartben commented:
@aescolar @nashif FWIW @nordicjm's request for change has been applied verbatim, in case he hasn't time to revisit before RC is cut

kartben commented:
> fwiw I was able to get blinky running on [docs.m5stack.com/en/core/M5NanoC6](https://docs.m5stack.com/en/core/M5NanoC6) 🌟 🚀 +1 for the docs aspects as well

https://github.com/zephyrproject-rtos/zephyr/pull/74697 in case anyone is interested :)
fwiw I cannot seem to get the UART to work, is this a known issue by any chance?
